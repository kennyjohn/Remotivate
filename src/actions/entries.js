import database from '../firebase/firebase';

// Action Creators
// ADD_ENTRY
export const addEntry = (entry) => ({
  type: 'ADD_ENTRY',
  entry
});

export const startAddEntry = (entryData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const { 
      text = '',
      author = '',
      video = '',
      image = '',
      purpose = '',
      createdAt = 0
    } = entryData;

    const entry = { text, author, video, image, purpose, createdAt };
    return database.ref(`users/${uid}/entries`).push(entry)
    .then((ref) => {
      dispatch(addEntry({
        id: ref.key,
        ...entry
      }));
    });
  };
};

// REMOVE_ENTRY
export const removeEntry = ({ id } = {}) => ({
    type: 'REMOVE_ENTRY',
    id
});

export const startRemoveEntry = ({ id } = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/entries/${id}`).remove()
        .then(() => dispatch(removeEntry({ id })));
    };
};

// EDIT_ENTRY
export const editEntry = (id, updates) => ({
    type: 'EDIT_ENTRY',
    id,
    updates
});

export const startEditEntry = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/entries/${id}`).update(updates)
        .then(() => dispatch(editEntry(id, updates)))
    };
}

export const setEntries = (entries) => ({
  type: 'SET_ENTRIES',
  entries
});

export const startSetEntries = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/entries`).once('value').then((snapshot) => {
      const entries = [];
      snapshot.forEach((childSnapshot) => {
        entries.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      dispatch(setEntries(entries));
      });
    });
  };
};