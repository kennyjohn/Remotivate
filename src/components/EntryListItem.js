import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

// https://stackoverflow.com/questions/25661182/embed-youtube-video-refused-to-display-in-a-frame-because-it-set-x-frame-opti

const EntryListItem = ({ id, text, author, video, image, purpose, createdAt }) => (
    <Link className={text ? "list-item onHoverText" : video || image ? "list-item onHoverMedia" : ""} to={`/edit/${id}`}>
    { image ? <img src={image}></img> :
        text ? <p>{text}</p> :
        video ? <iframe width="335" height="250" src={video.replace("watch?v=", "embed/")} frameBorder="0" allow="encrypted-media" allowFullScreen></iframe> :
        null
    }
    <p className="list-item__subtitle">{moment(createdAt).format('MMMM Do, YYYY')}</p>
    </Link>
);

export default EntryListItem;