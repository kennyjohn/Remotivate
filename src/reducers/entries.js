const entriesReducerDefaultState = [];

export default (state = entriesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_ENTRY':
            return [
                ...state,
                action.entry
            ];
        case 'REMOVE_ENTRY':
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_ENTRY':
            return state.map((entry) => {
                if(entry.id === action.id) {
                    return [
                        ...state,
                        ...action.updates
                    ];
                } else {
                    return entry;
                }
            });
        case 'SET_ENTRIES':
            return action.entries;
        default:
            return state;
    }
}