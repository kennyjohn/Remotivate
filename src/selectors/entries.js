import moment from 'moment';

export default (entries, { value, sortByDate, startDate, endDate }) => {
    return entries.filter((entry) => {
        const createdAtMoment = moment(entry.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textFilterMatch = entry.text ? true : false;
        const videoFilterMatch = entry.video ? true : false;
        const imageFilterMatch = entry.image ? true : false;
        if(value === "text") {
            return startDateMatch && endDateMatch && textFilterMatch;
        } else if (value === "video") {
            return startDateMatch && endDateMatch && videoFilterMatch;
        } else if (value === "image") {
            return startDateMatch && endDateMatch && imageFilterMatch;
        } else {
            return startDateMatch && endDateMatch;
        }
    }).sort((a, b) => {
      if (sortByDate) {
        return a.createdAt < b.createdAt ? 1 : -1;
      }
    });
};
  