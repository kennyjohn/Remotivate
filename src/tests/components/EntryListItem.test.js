import React from 'react';
import { shallow } from 'enzyme';
import entries from '../fixtures/entries';
import EntryListItem from '../../components/EntryListItem';

test('should render EntryListItem component', () => {
    const wrapper = shallow(<EntryListItem
        {...entries[1]}
    ></EntryListItem>)
    expect(wrapper).toMatchSnapshot();
});