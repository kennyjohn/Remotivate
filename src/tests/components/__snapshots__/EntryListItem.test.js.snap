// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EntryListItem component 1`] = `
<Link
  className="list-item onHoverMedia"
  replace={false}
  to="/edit/69312332223"
>
  <img
    src="https://wwww.pictureimage.com"
  />
  <p
    className="list-item__subtitle"
  >
    January 5th, 1970
  </p>
</Link>
`;
